File: internal/core/patch.go
package core

import "fmt"

type Config struct {
	Games map[string]Game `json:"games"`
}

type Game struct {
	Sites []Site `json:"sites"`
}

// Site is one patch recipe: locate Find (wildcards allowed), verify Expect at
// PatchOffset, write Replace. Optionally restrict to functions referencing Anchor
// and/or require the match to sit inside a real .pdata function.
type Site struct {
	ID            string `json:"id"`
	Anchor        string `json:"anchor,omitempty"`        // restrict search to funcs referencing this string
	Find          string `json:"find"`                    // wildcard byte pattern
	PatchOffset   int    `json:"patchOffset"`             // offset within a match to write Replace
	Expect        string `json:"expect"`                  // bytes that must be present before patching
	Replace       string `json:"replace"`                 // bytes to write (same length as Expect)
	RequireInFunc bool   `json:"requireInFunc,omitempty"` // match must be inside a .pdata function
	ExpectMatches int    `json:"expectMatches,omitempty"` // required match count (default 1); fail-closed otherwise
	Note          string `json:"note,omitempty"`
}

type Status string

const (
	StatusPatched Status = "patched"
	StatusAlready Status = "already-patched"
	StatusNoMatch Status = "not-found"
	StatusAmbig   Status = "ambiguous" // match count != expected; refused
	StatusError   Status = "error"
)

type SiteResult struct {
	Site    Site
	Status  Status
	Count   int      // matches of the unpatched pattern
	RVAs    []uint32 // patched locations
	Message string
}

// PatchImage applies every site into a clone of img.Raw. Transactional: if any site
// blocks (not-found/ambiguous/error) the clone is discarded and (results, nil, false)
// is returned, so a caller never writes a half-patched exe. img.Raw stays pristine.
func PatchImage(img *Image, game Game) ([]SiteResult, []byte, bool) {
	out := append([]byte(nil), img.Raw...)
	var results []SiteResult
	blocking, anyPatched := false, false
	for _, site := range game.Sites {
		r := applySite(img, out, site)
		switch r.Status {
		case StatusPatched:
			anyPatched = true
		case StatusAlready:
			// not a change, not a blocker
		default: // not-found/ambiguous/error: a required site failed, block the game
			blocking = true
		}
		results = append(results, r)
	}
	if blocking || !anyPatched {
		return results, nil, false
	}
	return results, out, true
}

// applySite locates the site's matches in img (pristine), and writes replacements
// into out (a same-length buffer, typically a clone of img.Raw). It never writes
// unless every match is bounds- and expect-verified first (two-phase).
func applySite(img *Image, out []byte, site Site) SiteResult {
	res := SiteResult{Site: site}
	want := site.ExpectMatches
	if want == 0 {
		want = 1
	}

	find, err := ParsePattern(site.Find)
	if err != nil {
		return res.err(err.Error())
	}
	expect, err := ParseBytes(site.Expect)
	if err != nil {
		return res.err("expect: " + err.Error())
	}
	replace, err := ParseBytes(site.Replace)
	if err != nil {
		return res.err("replace: " + err.Error())
	}
	if err := validateSite(site, find, expect, replace); err != nil {
		return res.err(err.Error())
	}

	text := img.Text()
	if text == nil {
		return res.err("no .text section")
	}
	tbytes := img.Bytes(text)
	if tbytes == nil {
		return res.err("unreadable .text section")
	}

	// Optional string-anchor: restrict to functions that reference the anchor.
	var anchorRanges []FnRange
	if site.Anchor != "" {
		anchorRanges = img.FuncsReferencingString(site.Anchor)
		if len(anchorRanges) == 0 {
			res.Status = StatusNoMatch
			res.Message = fmt.Sprintf("anchor %q not referenced in code", site.Anchor)
			return res
		}
	}

	keep := func(rva uint32) bool {
		if site.RequireInFunc || len(anchorRanges) > 0 {
			fr, in := img.FuncContains(rva)
			if !in {
				return false
			}
			if len(anchorRanges) > 0 && !containsRange(anchorRanges, fr) {
				return false
			}
		}
		return true
	}
	filter := func(offsets []int) []uint32 {
		var out []uint32
		for _, off := range offsets {
			rva := text.RVA + uint32(off)
			if keep(rva) {
				out = append(out, rva)
			}
		}
		return out
	}

	matches := filter(find.FindAll(tbytes))
	res.Count = len(matches)

	if len(matches) == 0 {
		// Distinguish already-patched from genuinely missing. Require the patched
		// pattern to appear exactly `want` times, else it's an odd/partial state.
		patched := filter(find.Overlay(site.PatchOffset, replace).FindAll(tbytes))
		switch {
		case len(patched) == want:
			res.Status = StatusAlready
		case len(patched) > 0:
			res.Status = StatusAmbig
			res.Message = fmt.Sprintf("found %d patched matches, expected %d (partial/foreign patch?)", len(patched), want)
		default:
			res.Status = StatusNoMatch
		}
		return res
	}
	if len(matches) != want {
		res.Status = StatusAmbig
		res.Message = fmt.Sprintf("found %d matches, expected %d", len(matches), want)
		return res
	}

	// Phase 1: resolve + verify every write before touching any bytes.
	type write struct {
		off int
		rva uint32
	}
	var writes []write
	for _, rva := range matches {
		off, ok := img.RVAToFileOff(rva + uint32(site.PatchOffset))
		if !ok || off+len(replace) > len(out) || off+len(expect) > len(img.Raw) {
			return res.err(fmt.Sprintf("write range out of bounds at RVA 0x%X", rva+uint32(site.PatchOffset)))
		}
		if !bytesEqual(img.Raw[off:off+len(expect)], expect) {
			return res.err(fmt.Sprintf("expect bytes mismatch at RVA 0x%X", rva+uint32(site.PatchOffset)))
		}
		writes = append(writes, write{off, rva})
	}
	// Phase 2: commit.
	for _, w := range writes {
		copy(out[w.off:w.off+len(replace)], replace)
		res.RVAs = append(res.RVAs, w.rva)
	}
	res.Status = StatusPatched
	return res
}

// validateSite rejects recipes whose patch would fall outside the located match —
// the main safeguard for untrusted override configs.
func validateSite(site Site, find Pattern, expect, replace []byte) error {
	if len(expect) == 0 || len(replace) == 0 {
		return fmt.Errorf("site %s: empty expect/replace", site.ID)
	}
	if len(expect) != len(replace) {
		return fmt.Errorf("site %s: expect (%d) and replace (%d) must be the same length", site.ID, len(expect), len(replace))
	}
	if site.PatchOffset < 0 || site.PatchOffset+len(expect) > len(find) {
		return fmt.Errorf("site %s: patchOffset+expect exceeds find pattern length", site.ID)
	}
	// The find pattern's concrete bytes at the patch offset must equal expect.
	for i := range expect {
		if b := find[site.PatchOffset+i]; b >= 0 && byte(b) != expect[i] {
			return fmt.Errorf("site %s: expect byte %d disagrees with find pattern", site.ID, i)
		}
	}
	return nil
}

func (r SiteResult) err(msg string) SiteResult {
	r.Status, r.Message = StatusError, msg
	return r
}

func containsRange(rs []FnRange, r FnRange) bool {
	for _, x := range rs {
		if x == r {
			return true
		}
	}
	return false
}

func bytesEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: src/core/patch.js
// Port of internal/core/patch.go

import { hexUpper, quote } from '../internal/gostrconv.js';
import { CoreError, findAll, overlay, parseBytes, parsePattern } from './pattern.js';
import { fnKey } from './image.js';

export const StatusPatched = 'patched';
export const StatusAlready = 'already-patched';
export const StatusNoMatch = 'not-found';
export const StatusAmbig = 'ambiguous';
export const StatusError = 'error';

export function bytesEqual(a, b) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

// patchImage applies every site into a clone of img.raw. Transactional: if any
// site blocks, the clone is discarded and nothing is returned to write.
export function patchImage(img, game) {
  const out = img.raw.slice();
  const results = [];
  let blocking = false;
  let anyPatched = false;
  for (const site of game.sites ?? []) {
    const r = applySite(img, out, site);
    if (r.status === StatusPatched) anyPatched = true;
    else if (r.status !== StatusAlready) blocking = true;
    results.push(r);
  }
  if (blocking || !anyPatched) return { results, patched: null, changed: false };
  return { results, patched: out, changed: true };
}

function errResult(res, msg) {
  res.status = StatusError;
  res.message = msg;
  return res;
}

export function applySite(img, out, site) {
  const res = { site, status: '', count: 0, rvas: [], message: '' };
  let want = site.expectMatches ?? 0;
  if (want === 0) want = 1;

  let find;
  let expect;
  let replace;
  try {
    find = parsePattern(site.find);
  } catch (err) {
    return errResult(res, err.message);
  }
  try {
    expect = parseBytes(site.expect);
  } catch (err) {
    return errResult(res, `expect: ${err.message}`);
  }
  try {
    replace = parseBytes(site.replace);
  } catch (err) {
    return errResult(res, `replace: ${err.message}`);
  }
  const vErr = validateSite(site, find, expect, replace);
  if (vErr !== null) return errResult(res, vErr);

  const text = img.text();
  if (text === null) return errResult(res, 'no .text section');
  const tbytes = img.bytes(text);
  if (tbytes === null) return errResult(res, 'unreadable .text section');

  let anchorRanges = [];
  if ((site.anchor ?? '') !== '') {
    anchorRanges = img.funcsReferencingString(site.anchor);
    if (anchorRanges.length === 0) {
      res.status = StatusNoMatch;
      res.message = `anchor ${quote(site.anchor)} not referenced in code`;
      return res;
    }
  }
  const anchorKeys = new Set(anchorRanges.map(fnKey));

  const keep = (rva) => {
    if (site.requireInFunc === true || anchorRanges.length > 0) {
      const { fr, in: inFunc } = img.funcContains(rva);
      if (!inFunc) return false;
      if (anchorRanges.length > 0 && !anchorKeys.has(fnKey(fr))) return false;
    }
    return true;
  };
  const filter = (offsets) => {
    const kept = [];
    for (const off of offsets) {
      const rva = (text.rva + off) >>> 0;
      if (keep(rva)) kept.push(rva);
    }
    return kept;
  };

  const matches = filter(findAll(find, tbytes));
  res.count = matches.length;

  if (matches.length === 0) {
    const patched = filter(findAll(overlay(find, site.patchOffset, replace), tbytes));
    if (patched.length === want) {
      res.status = StatusAlready;
    } else if (patched.length > 0) {
      res.status = StatusAmbig;
      res.message = `found ${patched.length} patched matches, expected ${want} (partial/foreign patch?)`;
    } else {
      res.status = StatusNoMatch;
    }
    return res;
  }
  if (matches.length !== want) {
    res.status = StatusAmbig;
    res.message = `found ${matches.length} matches, expected ${want}`;
    return res;
  }

  const writes = [];
  for (const rva of matches) {
    const target = (rva + site.patchOffset) >>> 0;
    const { off, ok } = img.rvaToFileOff(target);
    if (!ok || off + replace.length > out.length || off + expect.length > img.raw.length) {
      return errResult(res, `write range out of bounds at RVA 0x${hexUpper(target)}`);
    }
    if (!bytesEqual(img.raw.subarray(off, off + expect.length), expect)) {
      return errResult(res, `expect bytes mismatch at RVA 0x${hexUpper(target)}`);
    }
    writes.push({ off, rva });
  }
  for (const w of writes) {
    out.set(replace, w.off);
    res.rvas.push(w.rva);
  }
  res.status = StatusPatched;
  return res;
}

// validateSite rejects recipes whose patch would fall outside the located match.
// Returns null when valid, else the Go error text.
function validateSite(site, find, expect, replace) {
  if (expect.length === 0 || replace.length === 0) {
    return `site ${site.id}: empty expect/replace`;
  }
  if (expect.length !== replace.length) {
    return `site ${site.id}: expect (${expect.length}) and replace (${replace.length}) must be the same length`;
  }
  if (site.patchOffset < 0 || site.patchOffset + expect.length > find.length) {
    return `site ${site.id}: patchOffset+expect exceeds find pattern length`;
  }
  for (let i = 0; i < expect.length; i++) {
    const b = find[site.patchOffset + i];
    if (b >= 0 && b !== expect[i]) {
      return `site ${site.id}: expect byte ${i} disagrees with find pattern`;
    }
  }
  return null;
}
